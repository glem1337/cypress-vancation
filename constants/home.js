export const HERO_PAGE_CARDS = [
  {
    icon: 'Question',
    color: 'green',
    titleId: { id: 'homePage.heroPageCard.customerSupport' },
    descriptionId: { id: 'homePage.heroPageCard.stubDescription' },
  },
  {
    icon: 'Shield',
    color: 'yellow',
    titleId: { id: 'homePage.heroPageCard.fullyInsured' },
    descriptionId: { id: 'homePage.heroPageCard.stubDescription' },
  },
  {
    icon: 'Phone',
    color: 'red',
    titleId: { id: 'homePage.heroPageCard.includedRoadside' },
    descriptionId: { id: 'homePage.heroPageCard.stubDescription' },
  },
];

export const WORKS_CARDS = [
  {
    icon: 'Search',
    titleId: { id: 'homePage.works.findThePerfect' },
    descriptionId: { id: 'homePage.works.enterLocationAndDate' },
  },
  {
    icon: 'Calendar',
    titleId: { id: 'homePage.works.bookYourTrip' },
    descriptionId: { id: 'homePage.works.bookOnlineSecurely' },
  },
  {
    icon: 'Road',
    titleId: { id: 'homePage.works.hitTheRoad' },
    descriptionId: { id: 'homePage.works.haveYourDreamCampervan' },
  },
];

export const BROWSER_POPULAR = [
  {
    img: '/images/listing/Modern-Van.svg',
    link: '',
    titleId: { id: 'homePage.browserPopular.title.campervans' },
    descriptionId: { id: 'homePage.browserPopular.campervanRentalsAreAll' },
    buttonId: { id: 'homePage.browserPopular.browseCampervans' },
  },
  {
    img: '/images/listing/VW-Bus.svg',
    link: '',
    titleId: { id: 'homePage.browserPopular.title.vwBuses' },
    descriptionId: { id: 'homePage.browserPopular.westfaliasEurovans' },
    buttonId: { id: 'homePage.browserPopular.browseVWBuses' },
  },
  {
    img: '/images/listing/Unique-Camper.svg',
    link: '',
    titleId: { id: 'homePage.browserPopular.title.uniqueCampers' },
    descriptionId: { id: 'homePage.browserPopular.ifYouAreLookingFor' },
    buttonId: { id: 'homePage.browserPopular.browseUniqueCampers' },
  },
  {
    img: '/images/listing/Vehicle-Camper.svg',
    link: '',
    titleId: { id: 'homePage.browserPopular.title.vehicleCamper' },
    descriptionId: { id: 'homePage.browserPopular.doNotNeedLuxury' },
    buttonId: { id: 'homePage.browserPopular.browseVehicleCampers' },
  },
];

export const FEATURED_LOCATION_SECTION = {
  FETCH_FIRST_ITEMS: 20,
  SLIDE_ITEMS: 4,
  SLIDE_ITEMS_MOBILE: 1,
};

export const NEXT_TRIP_DEFAULT_SHOW_ITEMS = 24;
