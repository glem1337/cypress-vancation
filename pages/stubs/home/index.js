export { default } from 'views/stubs/home';
