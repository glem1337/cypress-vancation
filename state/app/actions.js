import {
  SET_PAGES,
  SET_CURRENT_LOCATION,
  SET_CURRENT_COORDINATES,
  SET_OPEN_GRAPH_DATA,
} from './type';

export const setPages = ({ currentPage, prevPage, isServer }) => ({
  type: SET_PAGES,
  isServer: isServer || false,
  currentPage,
  prevPage,
});

export const setCurrentCoordinates = ({ longitude, latitude }) => ({
  type: SET_CURRENT_COORDINATES,
  longitude,
  latitude,
});

export const setCurrentLocation = (location) => ({
  type: SET_CURRENT_LOCATION,
  location,
});

export const setOpenGraphData = (data) => ({
  type: SET_OPEN_GRAPH_DATA,
  data,
});
