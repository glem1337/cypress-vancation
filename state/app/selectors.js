import * as R from 'ramda';

export const prevPageSelector = R.path(['app', 'prevPage']);
export const currentPageSelector = R.path(['app', 'currentPage']);
export const currentCoordinatesSelector = R.path(['app', 'currentCoordinates']);
export const currentLocationSelector = R.path(['app', 'currentLocation']);
export const openGraphDataSelector = R.path(['app', 'openGraph']);
