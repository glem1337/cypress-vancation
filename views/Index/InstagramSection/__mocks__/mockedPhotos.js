export default [
  {
    id: 1,
    postUrl: 'postUrl',
    imageUrl: 'imageUrl',
  },
  {
    id: 2,
    postUrl: 'postUrl',
    imageUrl: 'imageUrl',
  },
  {
    id: 3,
    postUrl: 'postUrl',
    imageUrl: 'imageUrl',
  },
  {
    id: 4,
    postUrl: 'postUrl',
    imageUrl: 'imageUrl',
  },
  {
    id: 5,
    postUrl: 'postUrl',
    imageUrl: 'imageUrl',
  },
];
