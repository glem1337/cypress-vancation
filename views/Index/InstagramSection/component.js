import React from 'react';
import PropTypes from 'prop-types';
import { Row, Col, Button, Spin } from 'antd';
import { Swiper, SwiperSlide } from 'swiper/react';
import classNames from 'classnames';
import { FormattedMessage } from 'react-intl';
import SwiperCore, { Pagination } from 'swiper';
import { LoadingOutlined } from '@ant-design/icons';

import { INSTAGRAM, INSTAGRAM_USER_NAME } from 'constants/social-links';
import isPresent from 'utils/isPresent';

SwiperCore.use([Pagination]);

const InstagramSectionComponent = ({
  photos,
  swiperProps,
  swiperLayoutKey,
  moveLeft,
  moveRight,
  swiperRef,
  showPagination,
  onSwiperInit,
  canMoveLeft,
  onReachEnd,
  sliderWrapperRef,
  containerRef,
}) => (
  <div className="home-inner-wrap" ref={containerRef}>
    {isPresent(photos) && (
      <div className="container">
        <h2 className="home-title-sec mb-8">
          <FormattedMessage id="homePage.instagram.title" />
        </h2>
        <p className="text-title mb-16 font-400">
          <FormattedMessage id="homePage.instagram.subTitle" />
        </p>
        <div className="mb-40 mb-md-60">
          <Row>
            <Col span={16}>
              <a
                rel="noreferrer"
                className="d-inline-flex align-items-center"
                href={INSTAGRAM}
                target="_blank"
              >
                <div className="home-insta-link">
                  <i className="icon icon-instagram in-white font-20" />
                </div>
                <p className="text-title font-400">
                  {INSTAGRAM_USER_NAME}
                </p>
              </a>
            </Col>
          </Row>
        </div>
        <div className="main-slider-wrap" ref={sliderWrapperRef}>
          <Swiper
            className="home-slider-pos"
            key={swiperLayoutKey}
            ref={swiperRef}
            {...swiperProps}
              // onSlideChange={onSlideChangeHandler}
            onInit={onSwiperInit}
            watchSlidesProgress
            onReachEnd={onReachEnd}
          >
            {photos.map((photo) => (
              <SwiperSlide key={photo.id}>
                <a
                  rel="noreferrer"
                  href={photo.postUrl}
                  target="_blank"
                >
                  <div className="home-insta__img">
                    <Spin
                      indicator={<LoadingOutlined spin />}
                      className="search-page__spin"
                    />
                    <img src={photo.imageUrl} alt="" />
                  </div>
                </a>
              </SwiperSlide>
              ))}
          </Swiper>
          <div className={classNames('search-page__slider-nav', { 'd-none': !showPagination })}>
            <Button
              icon={<i className="icon icon-left font-14" />}
              type="secondary"
              shape="circle"
              onClick={moveLeft}
              disabled={!canMoveLeft}
            />
            <Button
              className="ml-16"
              icon={<i className="icon icon-right font-14" />}
              type="secondary"
              shape="circle"
              onClick={moveRight}
            />
          </div>
        </div>
      </div>
    )}
  </div>
  );

InstagramSectionComponent.propTypes = {
  photos: PropTypes.arrayOf(PropTypes.shape()),
  swiperProps: PropTypes.shape().isRequired,
  moveLeft: PropTypes.func.isRequired,
  moveRight: PropTypes.func.isRequired,
  swiperRef: PropTypes.shape().isRequired,
  swiperLayoutKey: PropTypes.string.isRequired,
  showPagination: PropTypes.bool.isRequired,
  onReachEnd: PropTypes.func.isRequired,
  onSwiperInit: PropTypes.func.isRequired,
  canMoveLeft: PropTypes.bool.isRequired,
  sliderWrapperRef: PropTypes.shape().isRequired,
  containerRef: PropTypes.shape().isRequired,
};

InstagramSectionComponent.defaultProps = {
  photos: [],
};

export default InstagramSectionComponent;
