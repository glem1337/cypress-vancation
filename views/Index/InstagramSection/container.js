import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { v4 as uuid } from 'uuid';
import deepEqual from 'fast-deep-equal';
import * as R from 'ramda';

import { fetchInstagramPhotos as fetchInstagramPhotosAction } from 'state/concepts/home/actions';
import {
  instagramPhotosSelector,
  instagramPhotosPageSelector,
  instagramPhotosTotalSelector,
} from 'state/concepts/home/selectors';
import { fetchInstagramPhotosEndpoint } from 'state/concepts/home/endpoints';
import { dataDeleteEntity as dataDeleteEntityAction } from 'state/data/actions';
import { loadingSelector } from 'state/data/selectors';
import isMobileView from 'utils/breakpoints/isMobileView';
import isTabletView from 'utils/breakpoints/isTabletView';
import isPresent from 'utils/isPresent';

import InstagramSectionComponent from './component';

const DESKTOP_SLIDES_COUNT = 4;
const INTERSECTION_OFFSET = 700;

class InstagramSection extends React.Component {
  static propTypes = {
    fetchInstagramPhotos: PropTypes.func.isRequired,
    dataDeleteEntity: PropTypes.func.isRequired,
    total: PropTypes.number.isRequired,
    page: PropTypes.number.isRequired,
    arePhotosFetching: PropTypes.bool,
    photos: PropTypes.arrayOf(PropTypes.shape()),
  }

  static defaultProps = {
    arePhotosFetching: false,
    photos: [],
  }

  constructor(props) {
    super(props);

    this.isDataRequested = false;

    this.activeSlideIndex = null;

    this.swiperRef = React.createRef();

    this.sliderWrapperRef = React.createRef();

    this.containerRef = React.createRef();

    this.state = {
      swiperProps: {
        watchOverflow: true,
        loop: props.total === props.photos?.length,
        loopFillGroupWithBlank: true,
        slidesPerView: DESKTOP_SLIDES_COUNT,
        slidesPerGroup: DESKTOP_SLIDES_COUNT,
        spaceBetween: 24,
        lazy: {
          loadPrevNext: true,
        },
        pagination: {
          type: 'fraction',
        },
      },
      isMobileOptions: undefined,
      showPagination: true,
      swiperLayoutKey: uuid(),
      allowTouchMove: true,
    };
  }

  /**
   * Lifecycle method.
   */
  componentDidMount() {
    window.addEventListener('resize', this.handleResize);

    this.observer = new IntersectionObserver(this.handleIntersect, {
      rootMargin: `${INTERSECTION_OFFSET}px`,
    });
    this.observer.observe(this.containerRef.current);
  }

  /**
   * Lifecycle method.
   */
  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
    this.observer.unobserve(this.containerRef.current);
  }

  /**
   * Lifecycle method.
   */
  componentDidUpdate(prevProps, prevState) {
    const { arePhotosFetching, photos } = this.props;

    if (!deepEqual(prevState, this.state) && this.swiperRef?.current?.swiper?.update) {
      this.swiperRef.current.swiper.update();
    }

    if (prevProps.arePhotosFetching && !arePhotosFetching) {
      if (this.sliderWrapperRef.current) {
        const spins = document.querySelectorAll('.home-insta__img .search-page__spin');
        // eslint-disable-next-line no-param-reassign
        spins.forEach(item => { item.style.zIndex = 10; });

        this.sliderWrapperRef.current.classList.remove('instagram_slider_wrapper--disabled');
      }

      if (this.swiperRef.current?.swiper?.update) {
        this.swiperRef.current?.swiper.update();
      }

      if (isPresent(photos)) {
        this.isDataRequested = true;
      }

      this.handleResize();
      this.createSwiperLoop();
    }
  }

  /**
   * Handle intersect.
   */
  handleIntersect = (entries) => {
    const { fetchInstagramPhotos } = this.props;

    const isIntersecting = R.path(['0', 'isIntersecting'], entries);

    if (isIntersecting && !this.isDataRequested) {
      fetchInstagramPhotos();
    }
  }

  /**
   * Handle resize event.
   */
   handleResize = () => {
    const { photos, total } = this.props;

    const { isMobileOptions } = this.state;

    const isMobile = isMobileView();
    const isTablet = isTabletView();
    const isDesktop = !isMobile && !isTablet;

    if ((isTablet || isMobile) && isMobileOptions !== true) {
      this.setState({
        swiperProps: {
          slidesPerView: 'auto',
          watchOverflow: true,
          freeMode: true,
          freeModeSticky: true,
          loop: total === photos.length,
          lazy: {
            loadPrevNext: true,
          },
          pagination: {
            type: 'fraction',
          },
        },
        isMobileOptions: true,
        showPagination: false,
        swiperLayoutKey: uuid(),
      });

      return;
    }

    if (isDesktop && isMobileOptions !== false) {
      this.setState({
        swiperProps: {
          watchOverflow: true,
          loop: total === photos.length,
          loopFillGroupWithBlank: true,
          slidesPerView: DESKTOP_SLIDES_COUNT,
          slidesPerGroup: DESKTOP_SLIDES_COUNT,
          spaceBetween: 24,
          lazy: {
            loadPrevNext: true,
          },
          pagination: {
            type: 'fraction',
          },
        },
        isMobileOptions: false,
        showPagination: photos?.length > DESKTOP_SLIDES_COUNT,
        swiperLayoutKey: uuid(),
      });
    }
  }

  /**
   * Move swiper left
   */
  moveLeft = () => {
    if (!this.swiperRef?.current?.swiper?.slidePrev) {
      return false;
    }

    this.swiperRef.current.swiper.slidePrev();

    return true;
  };

  /**
   * Move swiper right
   */
  moveRight = () => {
    if (!this.swiperRef?.current?.swiper?.slideNext) {
      return false;
    }

    this.swiperRef.current.swiper.slideNext();

    return true;
  }

  /**
   * Amount slides
   */
  get amountSlides() {
    const { total } = this.props;

    const { isMobileOptions } = this.state;

    const batch = isMobileOptions
      ? 1
      : DESKTOP_SLIDES_COUNT;

    return parseInt(total / batch, 10);
  }

  /**
   * On swiper init
   */
  onSwiperInit = (swiper) => {
    // eslint-disable-next-line no-param-reassign
    swiper.params.pagination.formatFractionTotal = () => this.amountSlides;
    swiper.pagination.update();
  }

  /**
   * Can move left
   */
  get canMoveLeft() {
    const { total, photos } = this.props;

    return photos.length === total;
  }

  /**
   * Create swiper loop
   */
  createSwiperLoop = () => {
    const { total, photos } = this.props;

    const { swiperProps } = this.state;

    if (!this.swiperRef.current?.swiper || total !== photos.length) {
      return false;
    }

    this.setState({
      swiperProps: {
        ...swiperProps,
        loop: true,
      },
    });

    return true;
  }

  /**
   * On reach end
   */
  onReachEnd = () => {
    const {
      photos,
      fetchInstagramPhotos,
      page,
      total,
    } = this.props;

    if (photos.length === total) {
      return false;
    }

    if (this.sliderWrapperRef.current) {
      this.sliderWrapperRef.current.classList.add('instagram_slider_wrapper--disabled');

      const spins = document.querySelectorAll('.home-insta__img .search-page__spin');
      // eslint-disable-next-line no-param-reassign
      spins.forEach(item => { item.style.zIndex = 12; });
    }

    fetchInstagramPhotos({ page: page + 1 });

    return true;
  }

  render() {
    return (
      <InstagramSectionComponent
        {...this.props}
        {...this.state}
        moveLeft={this.moveLeft}
        moveRight={this.moveRight}
        swiperRef={this.swiperRef}
        onSwiperInit={this.onSwiperInit}
        canMoveLeft={this.canMoveLeft}
        onReachEnd={this.onReachEnd}
        sliderWrapperRef={this.sliderWrapperRef}
        containerRef={this.containerRef}
      />
    );
  }
}

const mapStateToProps = (state) => ({
  photos: instagramPhotosSelector(state),
  page: instagramPhotosPageSelector(state),
  total: instagramPhotosTotalSelector(state),
  arePhotosFetching: loadingSelector(state, fetchInstagramPhotosEndpoint.endpoint),
});

const mapDispatchToProps = {
  fetchInstagramPhotos: fetchInstagramPhotosAction,
  dataDeleteEntity: dataDeleteEntityAction,
};

export { InstagramSection as InstagramSectionContainer };
export default connect(mapStateToProps, mapDispatchToProps)(InstagramSection);
