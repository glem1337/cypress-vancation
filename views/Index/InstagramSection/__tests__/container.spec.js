import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import diveTo from 'utils/testHelpers/diveToEnzyme';
import isMobileView from 'utils/breakpoints/isMobileView';
import isTabletView from 'utils/breakpoints/isTabletView';
import { instagramPhotosTotalSelector } from 'state/concepts/home/selectors';
import { fetchInstagramPhotos } from 'state/concepts/home/actions';

import InstagramSection, { InstagramSectionContainer } from '../container';
import mockedPhotos from '../__mocks__/mockedPhotos';

jest.mock('state/data/selectors', () => ({
  loadingSelector: jest.fn(() => false),
}));

jest.mock('utils/breakpoints/isMobileView', () => jest.fn());
jest.mock('utils/breakpoints/isTabletView', () => jest.fn());

const mockedTotal = 20;
jest.mock('state/concepts/home/selectors', () => ({
  instagramPhotosSelector: jest.fn(() => mockedPhotos),
  instagramPhotosPageSelector: jest.fn(() => 2),
  instagramPhotosTotalSelector: jest.fn(() => mockedTotal),
}));

jest.mock('uuid', () => ({
  v4: jest.fn(() => 'uuid/v4'),
}));

const observe = jest.fn();
const unobserve = jest.fn();

window.IntersectionObserver = jest.fn(() => ({
  observe,
  unobserve,
}));

const layoutContainer = (props) => {
  const wrapper = shallow(<InstagramSection {...props} />, { disableLifecycleMethods: true });
  const container = diveTo(wrapper, InstagramSectionContainer);
  const instance = container.instance();
  const setStateSpy = jest.spyOn(instance, 'setState');

  instance.swiperRef = {
    current: {
      swiper: {
        update: jest.fn(),
      },
    },
  };

  instance.sliderWrapperRef = {
    current: {
      classList: {
        add: jest.fn(),
        remove: jest.fn(),
      },
    },
  };

  instance.containerRef = { current: {} };

  instance.observer = {
    observe,
    unobserve,
  };

  return {
    wrapper,
    container,
    instance,
    setStateSpy,
  };
};

describe('InstagramSection container tests', () => {
  Object.defineProperty(
    window,
    'addEventListener',
    { value: jest.fn() },
  );

  Object.defineProperty(
    window,
    'removeEventListener',
    { value: jest.fn() },
  );

  Object.defineProperty(
    document,
    'querySelectorAll',
    { value: jest.fn(() => [{ style: {} }]) },
  );

  const store = configureStore()({});
  store.dispatch = jest.fn();

  const props = {
    store,
  };

  let container = null;
  let instance = null;
  let setStateSpy = null;

  beforeEach(() => {
    ({
      container,
      instance,
      setStateSpy,
    } = layoutContainer(props));

    const div = document.createElement('div');
    div.classList.add('home-insta__img');

    const spin = document.createElement('div');
    div.classList.add('search-page__spin');

    div.appendChild(spin);
    document.body.appendChild(div);

    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    expect(container).toMatchSnapshot();
  });

  it('checks `componentDidMount` instance method', () => {
    instance.componentDidMount();

    expect(window.addEventListener).toHaveBeenCalledWith('resize', instance.handleResize);

    expect(observe).toHaveBeenCalledWith(instance.containerRef.current);
  });

  it('checks `componentWillUnmount` instance method', () => {
    instance.componentWillUnmount();

    expect(window.removeEventListener).toHaveBeenCalledWith('resize', instance.handleResize);

    expect(unobserve).toHaveBeenCalledWith(instance.containerRef.current);
  });

  describe('checks `handleIntersect` instance method', () => {
    it('should fetch data', () => {
      instance.handleIntersect([{ isIntersecting: true }]);

      expect(store.dispatch).toHaveBeenCalledWith(fetchInstagramPhotos());
    });

    it('should not fetch data', () => {
      instance.handleIntersect([{ isIntersecting: false }]);

      expect(store.dispatch).not.toHaveBeenCalled();
    });
  });

  describe('checks `componentDidUpdate` instance method', () => {
    it('when photos were loaded', () => {
      const handleResizeSpy = jest.spyOn(instance, 'handleResize');
      const createSwiperLoopSpy = jest.spyOn(instance, 'createSwiperLoop');

      instance.componentDidUpdate({ arePhotosFetching: true }, instance.state);

      expect(handleResizeSpy).toHaveBeenCalled();
      expect(createSwiperLoopSpy).toHaveBeenCalled();
    });

    it('when photos were loaded and without refs', () => {
      const handleResizeSpy = jest.spyOn(instance, 'handleResize');
      const createSwiperLoopSpy = jest.spyOn(instance, 'createSwiperLoop');

      instance.swiperRef = {
        current: {
          swiper: {},
        },
      };

      instance.sliderWrapperRef = {
        current: null,
      };

      instance.componentDidUpdate({ arePhotosFetching: true }, instance.state);

      expect(handleResizeSpy).toHaveBeenCalled();
      expect(createSwiperLoopSpy).toHaveBeenCalled();
    });

    it('when component state was changed', () => {
      instance.componentDidUpdate(props, {
        ...instance.state,
        swiperLayoutKey: 'test',
      });

      expect(instance.swiperRef.current.swiper.update).toHaveBeenCalled();
    });
  });

  describe('checks `handleResize` instance method', () => {
    it('for mobile', () => {
      isMobileView.mockReturnValueOnce(true);
      isTabletView.mockReturnValueOnce(true);

      ({
        container,
        instance,
        setStateSpy,
      } = layoutContainer(props));

      container.setState({ isMobileOptions: false });

      instance.handleResize();

      expect(setStateSpy).toHaveBeenCalledWith({
        swiperProps: {
          slidesPerView: 'auto',
          watchOverflow: true,
          freeMode: true,
          freeModeSticky: true,
          loop: false,
          lazy: {
            loadPrevNext: true,
          },
          pagination: {
            type: 'fraction',
          },
        },
        isMobileOptions: true,
        showPagination: false,
        swiperLayoutKey: 'uuid/v4',
      });
    });

    it('for desktop', () => {
      isMobileView.mockReturnValueOnce(false);
      isTabletView.mockReturnValueOnce(false);

      ({
        container,
        instance,
        setStateSpy,
      } = layoutContainer(props));

      container.setState({ isMobileOptions: true });

      instance.handleResize();

      expect(setStateSpy).toHaveBeenCalledWith({
        swiperProps: {
          watchOverflow: true,
          loop: mockedTotal === mockedPhotos.length,
          loopFillGroupWithBlank: true,
          slidesPerView: 4,
          slidesPerGroup: 4,
          spaceBetween: 24,
          lazy: {
            loadPrevNext: true,
          },
          pagination: {
            type: 'fraction',
          },
        },
        isMobileOptions: false,
        showPagination: mockedPhotos?.length > 4,
        swiperLayoutKey: 'uuid/v4',
      });
    });

    it('for desktop and when isMobileOptions equals false', () => {
      isMobileView.mockReturnValueOnce(false);
      isTabletView.mockReturnValueOnce(false);

      ({
        container,
        instance,
        setStateSpy,
      } = layoutContainer(props));

      container.setState({ isMobileOptions: false });

      setStateSpy.mockClear();

      instance.handleResize();

      expect(setStateSpy).not.toHaveBeenCalled();
    });
  });

  describe('checks `moveLeft` method', () => {
    it('when swiper ref does not exist', () => {
      const retVal = instance.moveLeft();

      expect(retVal).toBe(false);
    });

    it('move to prev slide', () => {
      instance.swiperRef = {
        current: {
          swiper: {
            slidePrev: jest.fn(),
          },
        },
      };

      const retVal = instance.moveLeft();

      expect(retVal).toBe(true);
      expect(instance.swiperRef.current.swiper.slidePrev).toHaveBeenCalled();
    });
  });

  describe('checks `moveRight` method', () => {
    it('when swiper ref does not exist', () => {
      const retVal = instance.moveRight();

      expect(retVal).toBe(false);
    });

    it('move to prev slide', () => {
      instance.swiperRef = {
        current: {
          swiper: {
            slideNext: jest.fn(),
          },
        },
      };

      const retVal = instance.moveRight();

      expect(retVal).toBe(true);
      expect(instance.swiperRef.current.swiper.slideNext).toHaveBeenCalled();
    });
  });

  describe('checks `amountSlides` instance getter', () => {
    it('for mobile', () => {
      container.setState({ isMobileOptions: true });

      const { amountSlides } = instance;

      expect(amountSlides).toBe(20);
    });

    it('for desktop', () => {
      container.setState({ isMobileOptions: false });

      const { amountSlides } = instance;

      expect(amountSlides).toBe(5);
    });
  });

  it('checks `onSwiperInit` instance method', () => {
    const swiper = {
      params: {
        pagination: {
          formatFractionTotal: null,
        },
      },
      pagination: {
        update: jest.fn(),
      },
    };

    instance.onSwiperInit(swiper);

    expect(swiper.params.pagination.formatFractionTotal()).toBe(instance.amountSlides);
    expect(swiper.pagination.update).toHaveBeenCalled();
  });

  describe('checks `canMoveLeft` instance getter', () => {
    it('should return false', () => {
      const { canMoveLeft } = instance;

      expect(canMoveLeft).toBe(false);
    });

    it('should return true', () => {
      instagramPhotosTotalSelector.mockReturnValueOnce(5);

      ({ instance } = layoutContainer(props));

      const { canMoveLeft } = instance;

      expect(canMoveLeft).toBe(true);
    });
  });

  describe('checks `createSwiperLoop` instance getter', () => {
    it('should return false', () => {
      const res = instance.createSwiperLoop();

      expect(res).toBe(false);
    });

    it('should return true', () => {
      instagramPhotosTotalSelector.mockReturnValueOnce(5);

      ({ instance, setStateSpy } = layoutContainer(props));

      const res = instance.createSwiperLoop();

      expect(setStateSpy).toHaveBeenCalledWith({
        swiperProps: {
          ...instance.state.swiperProps,
          loop: true,
        },
      });

      expect(res).toBe(true);
    });
  });

  describe('checks `onReachEnd` instance getter', () => {
    it('should return false', () => {
      instagramPhotosTotalSelector.mockReturnValueOnce(5);

      ({ instance } = layoutContainer(props));

      const res = instance.onReachEnd();

      expect(res).toBe(false);
    });

    it('should return true', () => {
      const res = instance.onReachEnd();

      expect(store.dispatch).toHaveBeenCalledWith(fetchInstagramPhotos({ page: 3 }));

      expect(res).toBe(true);
    });

    it('should return true without refs', () => {
      instance.swiperRef = {
        current: {
          swiper: {},
        },
      };

      instance.sliderWrapperRef = {
        current: null,
      };

      const res = instance.onReachEnd();

      expect(store.dispatch).toHaveBeenCalledWith(fetchInstagramPhotos({ page: 3 }));

      expect(res).toBe(true);
    });
  });
});
