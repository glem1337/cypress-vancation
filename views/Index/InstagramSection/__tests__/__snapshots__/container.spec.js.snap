// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstagramSection container tests matches snapshot 1`] = `
<InstagramSectionComponent
  allowTouchMove={true}
  arePhotosFetching={false}
  canMoveLeft={false}
  containerRef={
    Object {
      "current": null,
    }
  }
  dataDeleteEntity={[Function]}
  fetchInstagramPhotos={[Function]}
  moveLeft={[Function]}
  moveRight={[Function]}
  onReachEnd={[Function]}
  onSwiperInit={[Function]}
  page={2}
  photos={
    Array [
      Object {
        "id": 1,
        "imageUrl": "imageUrl",
        "postUrl": "postUrl",
      },
      Object {
        "id": 2,
        "imageUrl": "imageUrl",
        "postUrl": "postUrl",
      },
      Object {
        "id": 3,
        "imageUrl": "imageUrl",
        "postUrl": "postUrl",
      },
      Object {
        "id": 4,
        "imageUrl": "imageUrl",
        "postUrl": "postUrl",
      },
      Object {
        "id": 5,
        "imageUrl": "imageUrl",
        "postUrl": "postUrl",
      },
    ]
  }
  showPagination={true}
  sliderWrapperRef={
    Object {
      "current": null,
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  swiperLayoutKey="uuid/v4"
  swiperProps={
    Object {
      "lazy": Object {
        "loadPrevNext": true,
      },
      "loop": false,
      "loopFillGroupWithBlank": true,
      "pagination": Object {
        "type": "fraction",
      },
      "slidesPerGroup": 4,
      "slidesPerView": 4,
      "spaceBetween": 24,
      "watchOverflow": true,
    }
  }
  swiperRef={
    Object {
      "current": null,
    }
  }
  total={20}
/>
`;
